ui/src/db-tables/WorkflowsTable.ts

```diff
@@ -0,0 +1,179 @@
+import { Workflow } from "../types/dbTypes";
+import { generateUniqueName, sortFileItem } from "../utils";
+import { v4 as uuidv4 } from "uuid";
+import { TableBase } from "./TableBase";
+import { indexdb } from "./indexdb";
+import {
+  deleteJsonFileMyWorkflows,
+  saveJsonFileMyWorkflows,
+} from "./DiskFileUtils";
+import { ESortTypes, ImportWorkflow } from "../RecentFilesDrawer/types";
+
+export class WorkflowsTable extends TableBase<Workflow> {
+  static readonly TABLE_NAME = "workflows";
+
+  constructor() {
+    super("workflows");
+  }
+
+  static async load(): Promise<WorkflowsTable> {
+    const instance = new WorkflowsTable();
+    return instance;
+  }
+
+  public async createFlow({
+    json,
+    name,
+    parentFolderID,
+    tags,
+  }: {
+    json: string;
+    name?: string;
+    parentFolderID?: string;
+    tags?: string[];
+  }): Promise<Workflow> {
+    const newFlowName = await generateUniqueName(name);
+    const uuid = uuidv4(); // ⇨ '9b1deb4d-3b7d-4bad-9bdd-2b0d7b3dcb6d'
+    const time = Date.now();
+    const newWorkflow: Workflow = {
+      id: uuid,
+      name: newFlowName,
+      json,
+      parentFolderID: parentFolderID,
+      updateTime: time,
+      createTime: time,
+      tags: tags ?? [],
+    };
+    //add to IndexDB
+    await indexdb.workflows.add(newWorkflow);
+    // add to disk file db
+    this.saveDiskDB();
+    // add to my_workflows/
+    saveJsonFileMyWorkflows(newWorkflow);
+    return newWorkflow;
+  }
+
+  public async updateFlow(id: string, input: Partial<Workflow>) {
+    const before = await this.get(id);
+    if (before == null) {
+      return;
+    }
+    const after = {
+      ...before,
+      ...input,
+      id,
+    };
+    const beforeStr = JSON.stringify(before);
+    const afterStr = JSON.stringify(after);
+    if (beforeStr === afterStr) {
+      // no change detected
+      return;
+    }
+    const newWorkflow: Workflow = after;
+    // When modifying the associated tag or modifying the directory, updateTime is not modified.
+    const updateKey = Object.keys(input);
+    const isModifyingTagOrFolder =
+      updateKey.length === 1 &&
+      ["tags", "parentFolderID"].includes(updateKey[0]);
+    if (!isModifyingTagOrFolder) {
+      newWorkflow.updateTime = Date.now();
+      // update parent folder updateTime
+      if (newWorkflow.parentFolderID != null) {
+        await foldersTable?.update({
+          id: newWorkflow.parentFolderID,
+          updateTime: Date.now(),
+        });
+      }
+    }
+    //update indexdb
+    await indexdb.workflows.update(id, newWorkflow);
+    await this.saveDiskDB();
+    // save to my_workflows/
+    if (input.name !== null || input.parentFolderID !== null) {
+      // renamed file or moved file folder
+      await deleteJsonFileMyWorkflows(before);
+      await saveJsonFileMyWorkflows(after);
+      return;
+    }
+    if (input.json != null) {
+      await saveJsonFileMyWorkflows(after);
+    }
```

Comment:
in future, maybe we should split workflowTable.update() into updateName() updateJson()... to avoid so many conditional checks in it!